### PR TITLE
cli: GC old profiles

### LINKED ISSUE #14608: cli: cleanup old profiles
We currently allow periodic profiles to be dumped but don't clean up after ourselves, causing lots of files to be written and potentially filling up the disk.
We should delete old profiles after either some amount of time, or some number of profiles.

### PR DESCRIPTION
**Pull request by petermattis**
Fixes #14608

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/14626)

### CODE REVIEW

File: pkg/cli/start.go [hunk 1]
```diff
@@ -118,9 +119,53 @@ func initInsecure() error {
 	return nil
 }
 
+// maxSizePerProfile is the maximum total size in bytes for profiles per
+// profile type.
+var maxSizePerProfile = envutil.EnvOrDefaultInt64(
```
Comment (mberhault) [inline pkg/cli/start.go]: I think we should go for number of profiles, not size.
goroutines especially can get quite big.

File: pkg/cli/start.go [hunk 2]
```diff
@@ -118,9 +119,53 @@ func initInsecure() error {
...
+
+// gcProfiles removes old profiles matching the specified prefix when the sum
+// of newer profiles is larger than maxSize. Requires that the suffix used for
+// the profiles indicates age (e.g. by using a date/timestamp suffix) such that
+// sorting the filenames corresponds to ordering the profiles from oldest to
+// newest.
+func gcProfiles(prefix string, maxSize int64) {
+	paths, err := filepath.Glob(prefix + "*")
+	if err != nil {
+		log.Fatal(context.Background(), err)
```
Comment (mberhault) [inline pkg/cli/start.go]: fatal seems a bit drastic.

File: pkg/cli/start.go [hunk 3]
```diff
@@ -118,9 +119,53 @@ func initInsecure() error {
...
+var maxSizePerProfile = envutil.EnvOrDefaultInt64(
+	"COCKROACH_MAX_SIZE_PER_PROFILE", 100<<20 /* 100 MB */)
+
+// gcProfiles removes old profiles matching the specified prefix when the sum
+// of newer profiles is larger than maxSize. Requires that the suffix used for
+// the profiles indicates age (e.g. by using a date/timestamp suffix) such that
+// sorting the filenames corresponds to ordering the profiles from oldest to
+// newest.
+func gcProfiles(prefix string, maxSize int64) {
+	paths, err := filepath.Glob(prefix + "*")
```
Comment (mberhault) [inline pkg/cli/start.go]: nit: ioutil.ReadDir will returns a sorted list of `os.FileInfo`, doing the sorting and the stat for you. You can then call `filepath.Match` on it to check the glob.

**Comment by mberhault** [comment]
I guess I'll close #14623

**Comment by petermattis** [comment]
> I guess I'll close #14623

Doh! What happened there? I self-assigned #14608 while working on this PR last night. Looks like you then self-assigned that issue this morning and worked on your PR. Not sure why I didn't see your PR in my inbox this morning, though.

File: pkg/cli/start.go (hunk 1 above)
Comment (petermattis) [inline pkg/cli/start.go]: That goroutine profiles can get quite big sounds like an argument for limiting by size.

**Comment by mberhault** [comment]
no, it's my bad, I should have checked first. Anyway, yours has a test, so it takes precedence.

File: pkg/cli/start.go (hunk 3 above)
Comment (petermattis) [inline pkg/cli/start.go]: Ok. Changed.

File: pkg/cli/start.go (hunk 2 above)
Comment (petermattis) [inline pkg/cli/start.go]: Switched to a warning.

File: pkg/cli/start.go (hunk 1 above)
Comment (mberhault) [inline pkg/cli/start.go]: I think I was more concerned with having very little history left.

However, I remembered that we're not talking about goroutine profiles here, which are the ones that can get massive.
A big memory profile is still in the single-digit MB. I don't think cpu is too big either.

So yeah, let's stick with size-based.

**Comment by mberhault** [comment]
LGTM.

File: pkg/cli/start_test.go [hunk 4]
```diff
@@ -71,3 +78,49 @@ func TestInitInsecure(t *testing.T) {
...
+			t.Fatal(err)
+		}
+		sort.Strings(paths)
+		if e := expected[i-1:]; !reflect.DeepEqual(e, paths) {
+			t.Fatalf("%d: expected\n%s\nfound\n%s\n",
+				i, strings.Join(e, "\n"), strings.Join(paths, "\n"))
+		}
+		sum -= len(data[:i])
+	}
+}
```
Comment (spencerkimball) [inline pkg/cli/start_test.go]: Does it make sense to move these into a new source file? `start.go` risks becoming cluttered and these are really better located in some kind of debug or profiling grouping.

Comment (petermattis) [inline pkg/cli/start_test.go]: I'm comfortable with leaving this code here as only the `start` command uses it. I'm also fine with seeing it move to a new file, though I don't see the impetus for doing so right now.

**Pull request merged**